Depthaiv3 alpha15

phodina opened the pull request:
Add DepthAI package which is basically OpenCV running on Edge devices made by Luxonis.

Currently work in progress therefore marked as Draft, would like to decrease the amoutn of patches by upstreaming some of the changes like

- https://github.com/luxonis/depthai-core/pull/1303

[Package request](https://github.com/NixOS/nixpkgs/issues/292618) 

[Github repo link](https://github.com/luxonis/depthai-core)

Upstreaming the [flake environment](https://github.com/phodina/depthai-nix)

WIP Dependencies:
- https://github.com/NixOS/nixpkgs/pull/393017
- https://github.com/NixOS/nixpkgs/pull/393027
- https://github.com/NixOS/nixpkgs/pull/393020

CC: @jayrahdevore @tmayoff @hacker1024 @Pandapip1 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin (don't have any around, might fail on install of the install of shared library)
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

phodina committed:
depthai-core: init at 3.0.0-beta.1

phodina committed:
depthai-data: 0-unstable-2025-05-22

phodina committed:
ws-protocol: init at 1.4.0

phodina committed:
nanorpc: init at 1.1.0-unstable-2021-04-29

phodina committed:
mcap: init at 0.0.53

phodina commented:
deptahai-core:
- https://github.com/luxonis/depthai-core/pull/1309
- https://github.com/luxonis/depthai-core/pull/1302

nixpkgs:
- https://github.com/NixOS/nixpkgs/pull/403078

phodina commented:
- https://github.com/NixOS/nixpkgs/pull/403366

The `libnop` and `mp4v2` don't populate the `lib/cmake` so they are not found.

Porting `basalt` is still WIP, biggest issue is the large dependency on third party packages from ROS.

Pandapip1 commented on pkgs/by-name/ba/backward-cpp/package.nix:
```diff
@@ -2,33 +2,29 @@
-    maintainers = [ ];
+    license = lib.licenses.mit;
+    platforms = lib.platforms.all;
+    maintainers = with lib.maintainers; [ ];
```
Would you consider adding yourself as a maintainer?

Pandapip1 commented:
> Porting basalt is still WIP, biggest issue is the large dependency on third party packages from ROS.

I was working on ROS stuff at https://github.com/NixOS/nixpkgs/pull/355629, but it stalled because it was noted that ROS is pretty actively developed. If you want I can restart the effort.

phodina committed:
backward-cpp: add phodina as maintainer

phodina commented:
> I was working on ROS stuff at https://github.com/NixOS/nixpkgs/pull/355629, but it stalled because it was noted that ROS is pretty actively developed. If you want I can restart the effort.

AFAIK `basalt` is not required. It's optional dependency, though would love to make it work. However, first I'd focus just on DAI and then on the ROS ecosystem and finding way to package `basalt`.

Would like to do one more pass of the DAI package and mark the optional stuff and split package into multiple outputs.

Also it might be work to make two definitions - one that does not have dependencies on GUI stuff (good for containers) and then full blown implementation.

phodina commented:
FIY Not required as dependency but useful tool for OAK4 line

https://github.com/NixOS/nixpkgs/pull/403889

phodina commented:
Rebased and removed the already merged commits for `libnop` and `cpr`

Pandapip1 commented:
> Should I also squash the commits for depthai-data and depthai-core`?

Nope! These are different packages and should keep their separate commits.

phodina commented:
> Nope! These are different packages and should keep their separate commits. The individual commits for each package probably should be squashed into one commit per package, but this is no means a requirement. See https://github.com/NixOS/nixpkgs/issues/347586

Agreed on that.

Well meant just per package as each has several changes so that there's just one `depthai-core` and another for `depthai-data`. IMHO I don't see benefit in keeping both the `alpha14` and `alpha15` versions as there are alpha.

phodina commented:
There are additional points. IMHO I'd split the definition into multiple subpackages:
- `depthai-core-minimal` (OpenCV without GTK support, optional CMake flags turned off)
- `depthai-core` (OpenCV with GTK support, more CMake flags turned on)
- `depthai-core-full` (turn on examples and tests, for C++ and Python)

Also there are tests but they require HW attached or available on either the USB or network. Therefore they can't be just enabled and executed as part of the derivation, right? Or is there as it's similar to GPU as they are also not available in the sandbox environment but do have test suites.

If so could set up hardware to run the tests to see if everything is working (I just use subset of the DAI features that in my case so far work).

phodina committed:
pybind-11-mkdoc: init at 0-unstable-2023-02-08

phodina commented:
So pushed changes tracking the `v3_develop` branch as there were many changes that allow more CMake config options and more offline build.

Pandapip1 commented on pkgs/development/python-modules/pybind11-mkdoc/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  format = "pyproject";
+
+  meta = with lib; {
+    description = "Docstring extraction tool for pybind11 C++ bindings";
+    homepage = "https://github.com/pybind/pybind11_mkdoc";
+    license = licenses.mit;
+    maintainers = with maintainers; [ phodina ];
+    mainProgram = "pybind11_mkdoc";
+  };
```
Consider:
```suggestion
  meta = {
    description = "Docstring extraction tool for pybind11 C++ bindings";
    homepage = "https://github.com/pybind/pybind11_mkdoc";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ phodina ];
    mainProgram = "pybind11_mkdoc";
  };
```

Pandapip1 commented on pkgs/development/python-modules/pybind11-mkdoc/default.nix:
```diff
@@ -0,0 +1,36 @@
+{
+  lib,
+  python,
+  fetchFromGitHub,
+  buildPythonPackage,
+  flit-core,
+  clang,
+}:
+
+buildPythonPackage rec {
+  pname = "pybind11_mkdoc";
+  version = "0-unstable-2023-02-08";
+
+  src = fetchFromGitHub {
+    owner = "pybind";
+    repo = "pybind11_mkdoc";
+    rev = "42fbf377824185e255b06d68fa70f4efcd569e2d";
+    hash = "sha256-rXcJkOrxOGr1Ou+BuQYZzJTw8BV3h71/uYkpnMbP3Q0=";
+  };
+
+  nativeBuildInputs = [ flit-core ];
+  propagatedBuildInputs = [
+    python.pkgs.docutils
+    python.pkgs.libclang
```
Use this so that `.override` works:
```suggestion
  fetchFromGitHub,
  buildPythonPackage,
  flit-core,
  docutils,
  libclang,
  clang,
}:

buildPythonPackage rec {
  pname = "pybind11_mkdoc";
  version = "0-unstable-2023-02-08";

  src = fetchFromGitHub {
    owner = "pybind";
    repo = "pybind11_mkdoc";
    rev = "42fbf377824185e255b06d68fa70f4efcd569e2d";
    hash = "sha256-rXcJkOrxOGr1Ou+BuQYZzJTw8BV3h71/uYkpnMbP3Q0=";
  };

  nativeBuildInputs = [ flit-core ];
  propagatedBuildInputs = [
    docutils
    libclang
```

Pandapip1 commented on pkgs/servers/home-assistant/custom-components/solis-sensor/package.nix:
```diff
@@ -9,13 +9,13 @@
   owner = "hultenvp";
   domain = "solis";
-  version = "3.9.2";
+  version = "3.11.0";
```
Is there a reason this version bump is included in this PR?

phodina replied:
Just mistake during rebase.

Pandapip1 commented on pkgs/by-name/na/nanorpc/package.nix:
```diff
@@ -0,0 +1,39 @@
+  cmakeFlags = [
+    "-DBUILD_EXAMPLES=OFF" # Disable examples by default
+    "-DBUILD_TESTS=OFF" # Disable tests by default
+    "-DNANORPC_PURE_CORE=ON" # Skip Boost
+    "-DNANORPC_WITH_SSL=OFF" # Skip OpenSSL
```
Could you add a comment explaining why these build options are used?

Also, consider using `lib.cmakeBool`

Pandapip1 commented on pkgs/by-name/mc/mcap/package.nix:
```diff
@@ -0,0 +1,48 @@
+
+  sourceRoot = "source/cpp";
+
+  # Skip the build phase entirely
+  dontBuild = true;
+
+  # Skip the configure phase
+  dontConfigure = true;
```
This feels like the wrong order

```suggestion
  # Skip the configure phase
  dontConfigure = true;

  # Skip the build phase entirely
  dontBuild = true;
```

Pandapip1 commented on pkgs/by-name/de/depthai-data/package.nix:
```diff
@@ -0,0 +1,35 @@
+  fetchFromGitHub,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
```
Should this use `stdenvNoCC` instead?

Pandapip1 reviewed (commented):
I'll probably have more comments on a future pass by my battery is running out so here are some comments thus far.

phodina committed:
depthai-core: Split into depthai-core-minimal, depthai-core and depthai-core-full

phodina commented:
> There are additional points. IMHO I'd split the definition into multiple subpackages:
>
>    depthai-core-minimal (OpenCV without GTK support, optional CMake flags turned off)
>    depthai-core (OpenCV with GTK support, more CMake flags turned on)
>    depthai-core-full (turn on examples and tests, for C++ and Python)

Split the package into multiple based on enabled features.

> Also there are tests but they require HW attached or available on either the USB or network. Therefore they can't be just enabled and executed as part of the derivation, right? Or is there as it's similar to GPU as they are also not available in the sandbox environment but do have test suites.

Also the nix package builds as a part of `depthai-core` repo Github CI. The tests are next step to implement.

Lastly I'd like to also split further the outputs - probably `dev`, `out`, `python` and `share`.

phodina commented:
What needs to improve/minimize is the `pkgs/by-name/de/depthai-core/0001-nixos-specific-patches.patch`. The patched out lines are mostly due to VCPKG that has CMake configs for the dependencies but they are not present in Nix so needs revert to `pkg-config`.

phodina commented:
Any idea how to handle the check if the package is split?
```
- Because pkgs/by-name/de/depthai-core exists, the attribute `pkgs.depthai-core` must be defined like

    depthai-core = callPackage ./../by-name/de/depthai-core/package.nix { /* ... */ };

  However, in this PR, it isn't defined that way. See the definition in pkgs/top-level/all-packages.nix:363

    depthai-core
```

Pandapip1 reviewed (changes requested):
First pass. Please excuse any typos, this was written on my phone over a train ride.

Pandapip1 commented on pkgs/by-name/na/nanorpc/package.nix:
```diff
@@ -0,0 +1,37 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "nanorpc";
+  version = "1.1.0-unstable-2021-04-29";
```
The actual nanorpc package should probably be upstream's, since it now supports up to C++17 instead of C++14:

https://github.com/luxonis/nanorpc/compare/master..eb363c7004580d711f88ee1785d87ae6fc469220

Pandapip1 commented on pkgs/by-name/na/nanorpc/package.nix:
```diff
@@ -0,0 +1,37 @@
+  openssl,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
```
`finalAttrs` is seemingly not used anywhere here.

Pandapip1 commented on pkgs/by-name/de/depthai-core/generic.nix:
```diff
@@ -0,0 +1,315 @@
+        rm include/3rdparty -rf
+      '';
+
+  meta = {
```
Missing `maintainers`

Pandapip1 commented on pkgs/by-name/de/depthai-core/generic.nix:
```diff
@@ -0,0 +1,315 @@
+
+  prefixVar = "\${" + "prefix}";
+  # Latest commits are not compatible, use older version
+  xtensorCompat = xtensor.overrideAttrs (old: {
```
`old` is seemingly not used

```suggestion
  xtensorCompat = xtensor.overrideAttrs {
```

Pandapip1 commented on pkgs/by-name/de/depthai-core/generic.nix:
```diff
@@ -0,0 +1,315 @@
+  });
+
+  # Latest commits are not compatible, use older version
+  ws-protocolCompat = ws-protocol.overrideAttrs (old: {
```
`old` is seemingly not used

```suggestion
  ws-protocolCompat = ws-protocol.overrideAttrs {
```

Pandapip1 commented on pkgs/by-name/de/depthai-core/generic.nix:
```diff
@@ -0,0 +1,315 @@
+  });
+
+  # Latest commits are not compatible, use older version
+  xlinkCompat = xlink.overrideAttrs (old: {
```
`old` is seemingly not used

```suggestion
  xlinkCompat = xlink.overrideAttrs {
```

Pandapip1 reviewed (changes requested):
2nd pass
